### ISSUE TITLE
Incorrect values in screen variable when using if statements

### ISSUE THREAD

**Issue by redlesquire** [renpy/renpy#4163]
Here is an entire sample project
```
label start:
    call screen test()

screen test():
    vbox:
        $ x = "default"
        if True:
            text "in if"
            $ x = "if side"
        else:
            text "in else"
            $ x = "else side"
        text "[x]"
```
The output displayed is:
![Screenshot_20221217_125637](https://user-images.githubusercontent.com/120773555/208229787-e3ecbc8a-e321-489f-8adf-c6c09369a387.png)

If the code is changed so the if statement line reads `if False:` the results are still mixed up but flipped, ie:
in else
if side

**Comment by Gouvernathor** [comment]
Can't reproduce in 8.1.0
![image](https://user-images.githubusercontent.com/44340603/208240640-7d9514f3-d309-476a-b964-30e6ebc1b4e7.png)

**Comment by Gouvernathor** [comment]
If you manage to reproduce the bug, can you test what the following does ?
```rpy
screen pi():
    vbox:
        for k in range(10):
            text "[k]"
            $ kk = k
        if 1.0 == 1:
            text "True"
            $ txt = True
        else:
            text "False"
            $ txt = False
        text "[txt], [kk]"
```
For me, it does this, as intended.
![image](https://user-images.githubusercontent.com/44340603/208269603-f217ca8f-ad73-4eb4-8663-508bed67a689.png)

**Comment by redlesquire** [comment]
I'm running 8.0.3, which as far as I understand is the current official stable release.

I'll add that I'm running through the sdk launcher rather than a build.  If I run the project I get the results listed above.  If I reload the already running script in-progress then the display matches the expected output rather than the error state.

@Gouvernathor
Running your code instead I get 0-9 printed as expected followed by
True
False, 9

I will add that if I delete the else clause there can still be bleed-over.  With or without the added `$ txt - False` I get the same output.
```
screen pi():
    vbox:
        $ txt = False
        for k in range(10):
            text "[k]"
            $ kk = k
        if 1.0 == 2.0:
            text "True"
            $ txt = True
        # else:
        #     text "False"
        #     $ txt = False
        text "[txt], [kk]"
```
produces
0-9 printed one-per-line
True, 9

**Comment by KiloOscarSix** [comment]
@Gouvernathor I'm able to reproduce using renpy8.0.3.22090809

However, using `call screen` in the console returns the expected result.
Using `call screen` in the script gives the error, I'm guessing it's to do with screen prediction.

Adding `predict False` to the screen does produce the correct result using in-script `call` further affirming the issue lies in screen-prediction

**Comment by redlesquire** [comment]
Warning:  Conjecture to follow.  I avoided comments like this one in early posts for exactly that reason.

Yes, my expectation is also that it is the multi-pass predictive pass.  In some cases one side of the if statement is completely impossible to ever reach (as with my `if True` or `if False`.  The only way for it to be accessed is for something testing partial blocks of code irrespective of the entire flow.  Basically it is processing the code correctly on the real pass, but the "test" cases and real case are apparently sharing the same "scope level."  In other words, a "local variable" is a single copy being reused for every pass, real or fake rather than a truly local per-call variable.  It is basically the "official reason why global variables should not be modified, only read, in screens," but with local variables.  It is treating the screen variables as if they were global (as far as any single screen is concerned).  Which also may or may not make timing a factor.

**Comment by redlesquire** [comment]
@KiloOscarSix
As for switching $ to default... interesting.  So it sounds like "default" creates a true local variable.  "$" on the other hand, is creating a global variable (although possibly a "temporary global" that may be being cleaned up and deleted once the screen loses scope).

**Comment by redlesquire** [comment]
One last add...
```
screen pi():
    vbox:
        if False:
            text "True"
            $ txt = True
        text "[txt]"
```
The above code "should" produce an error.  A variable called "txt" should never have been defined when its value is read.  Yet it does not produce an error (it is filled with "True").  Likely it is being created and defined in the predictive loops.

**Comment by mal** [comment]
I believe this happens because you're actively fighting screen optimization as documented in https://www.renpy.org/doc/html/screen_optimization.html.

As it explains, `text "[blah]"` is analysed as treated as const, however you're actively manipulating the values you wish to interpolate inside branching code paths. If you must do this, then try using a non-const approach to using the manipulated value; i.e. a python expression such as `text blah`, `text f'{blah}'`, `text "%s' % blah` or `text '{}'.format(blah)`.

```rpy
screen pi():
    vbox:
        for k in range(10):
            text "[k]"
            $ kk = k

        if 1.0 == 1:
            text "True"
            $ txt = True
        else:
            text "False"
            $ txt = False

        text "concat: " + str(txt) + ", " + str(kk)      # concat:   True, 9
        text f"f-string {txt}, {kk}"                     # f-string: True, 9
        text "pctfmt %s, %s" % (txt, kk)                 # pctfmt:   True, 9
        text "format {}, {}".format(txt, kk)             # format:   True, 9
        text "const [txt], [kk]"                         # const:    False, 9
```

As such, I think this is broadly speaking, "expected" behaviour - at least within the context of how screen optimization works.

**Comment by redlesquire** [comment]
> I believe this happens because you're actively fighting screen optimization as documented in https://www.renpy.org/doc/html/screen_optimization.html.
> 
> As it explains, `text "[blah]"` is analysed as treated as const, however you're actively manipulating the values you wish to interpolate inside branching code paths. If you must do this, then try using a non-const approach to using the manipulated value; i.e. a python expression such as `text blah`, `text f'{blah}'`, `text "%s' % blah` or `text '{}'.format(blah)`.
> 
> ```python
> screen pi():
>     vbox:
>         for k in range(10):
>             text "[k]"
>             $ kk = k
> 
>         if 1.0 == 1:
>             text "True"
>             $ txt = True
>         else:
>             text "False"
>             $ txt = False
> 
>         text "concat: " + str(txt) + ", " + str(kk)      # concat:   True, 9
>         text f"f-string {txt}, {kk}"                     # f-string: True, 9
>         text "pctfmt %s, %s" % (txt, kk)                 # pctfmt:   True, 9
>         text "format {}, {}".format(txt, kk)             # format:   True, 9
>         text "const [txt], [kk]"                         # const:    False, 9
> ```
> 
> As such, I think this is broadly speaking, "expected" behaviour - at least within the context of how screen optimization works.
> 
> Ultimately screens are designed to be more declarative than procedural, so a better approach might be to have cleaner separation between your business logic and screen code; for example, providing read-only values to your screen and displaying based on those constant values which would also allow you to take full advantage of Ren'Py's optimisations in that space.

Just FYI, my original code where I found the problem *was not* using "text" at all.  This example project is reduced/minimalist creation to easily demonstrate the problem.  My original code was using existing variables which do not change for any given lifetime of the screen in question.  It was polling those variables to determine which specific image should be shown to the player.

Rather than writing a single rather complex (and therefore often difficult to follow) line, I was using a local, temporary, screen-scoped (or so I thought... actually I had even gone so far as to use the `renpy.dynamic()` function to make sure my variable was locally scoped and temporary) variable to make the code readable.

After finding that my variable sometimes had bad-data in it (the wrong image was showing up), I extracted my "if" statement into a full python block and assign my local-not local variable to the return value.  This allowed my code to work.

I'm new to ren'py, although not to code.  There are certainly lots of gotchas I'm not well versed in here.  But I'm not sure that saying, "you can use local variables and you can use if statements, but you can not use them together" is a valid end-game.

**Comment by mal** [comment]
Unfortunately it's not that straight forward. Using `$` is essentially running arbitrary python code, for which there are many valid applications within screens, however it's not feasible determine if any of those blocks perform assignment operations, and nor is forgoing the optimisations which help make screens as quick as they are,

**Comment by redlesquire** [comment]
I will add... according to the Screen Language Optimization documentation, 
`text "[txt]"` is considered a constant value as you say, but `text "%s" % txt` is considered "not const."  In the example code above changing the text line to a version that uses the %s substitution method also produces identical incorrect results.

**Comment by redlesquire** [comment]
Ah, interestingly I had initially only replaced the txt variable, as that is the one in debate (simply editing quickly).  I went back and did further testing and extracted both.  In other words, the text line had a mix of constant and non-constant values in my previous test (granted, to me that should result in being considered non-constant but appears to be considered constant instead).  `text "%s, [kk]" % txt`  That's a potential separate issue.  It does appear that if both variables are extracted it does evaluate correctly.

Unfortunately the original code with images does not exist.  I was toying around with concepts (which is part of why some of the logic was in a potentially less elegant place to begin with, not that that has any baring on what should or should not work) and it is all done differently now.  Maybe I will try to recreate.  Looking at the documentation, while it is all within the "text" displayable, it appears that it has nothing to do with the text displayable at all.  It is simply with the new-style substitution method wherever it is used.  It is certainly possible that my check was passing around string representations of images rather than actual image objects.  So it may boil down to the same string substitution scenario.

At a minimum I would suggest the documentation for string substitution should be modified to include a warning/strong note about this "feature."  The average ren'py target audience is not expected to deep-dive into screen optimization rules.  And the how-to in the text documentation portion suggests using the [] as the standard way to display variables in a screen with no mention of the fact the screen may substitute incorrect values in certain scenarios if this technique is used.

Not sure how
```
for i in range(3):
   text "[i]"
```
is considered "constant" while an "if" version is not.  Both are conditionally changing the value of the variable at some point within the screen prior to interpreting it into a string literal that is output.  (Not directly relevant, but somewhat inconsistently-consistent)

Still do not grasp why
```
default x = 0
if False:
    x = 1
text "[x]"
```
and
```
renpy.dynamic("x")
$ x = 0
if False:
    x = 1
text "[x]"
```
should work differently (whether the "dynamic" line is included or not).  Just tallying up points of note where two things described as doing the same thing yield different results.

**Comment by redlesquire** [comment]
tl;dr (although not short)
Screen Optimization Anomalies with Screen-Local Variables
1. screen local variables should all be declared with the `default` or `define` tag.  `$` declaration can result in optimization anomalies (known to exist when combined with if-statements and new-style `"[]"` string substitution)
2. new-style string substitution within screens (`"[myvar]"`) should only be used on global variables.  Local variables should use `%s` or other substitution methods (`[]` substitution can result in anomalies)
3. documentation only mentions `[]` string substitution method as being treated specially (as a constant) within the text-displayable screen optimization docs.  It is not related to the text-displayable at all, but any string substitution.   Suggestions: **Promote outside of text-displayable categorization.**  **Promote outside of optimization docs to general usage.**
4.  Combining constant and variable string substitution methods is treated as a constant!  `"[myvar1] and %s" % myvar2` (mixing is bad code style, but combining a constant and a variable should result in a variable classification)

**Comment by renpytom** [comment]
The first example looks like it should work, so I'm still treating this as an issue.

**Comment by Gouvernathor** [comment]
Despite the unclear or incorrect wording of the optimization page, `text "[a]"`, with `a` being non-const, shouldn't be expected to be const.
This syntax is optimized for const variables, that's fine, but it should not yield incorrect results with non-const values. Only take slower to compute, at worst.

**Comment by renpytom** [comment]
The problem is a bit more complicated here. Text should be considered const, but the scope update mechanism that updates the interpolations is failing here.
